theseus/optimizer/autograd/lu_cuda_sparse_autograd.py

```diff
@@ -46,22 +46,23 @@ def forward(  # type: ignore
 
         batch_size = A_val.shape[0]
 
+        A_val_double = A_val.double()
         AtA = mult_MtM(
-            batch_size, A_row_ptr, A_col_ind, A_val, AtA_row_ptr, AtA_col_ind
+            batch_size, A_row_ptr, A_col_ind, A_val_double, AtA_row_ptr, AtA_col_ind
         )
         if damping_alpha_beta is not None:
             AtA_args = sparse_structure.num_cols, AtA_row_ptr, AtA_col_ind, AtA
             apply_damping(batch_size, *AtA_args, *damping_alpha_beta)
         solver_context.factor(AtA)
 
-        A_args = sparse_structure.num_cols, A_row_ptr, A_col_ind, A_val
-        Atb = tmat_vec(batch_size, *A_args, b)
+        A_args = sparse_structure.num_cols, A_row_ptr, A_col_ind, A_val_double
+        Atb = tmat_vec(batch_size, *A_args, b.double())
         x = Atb.clone()
         solver_context.solve(x)  # solve in place
 
         ctx.b = b
         ctx.x = x
-        ctx.A_val = A_val
+        ctx.A_val_double = A_val_double
         ctx.A_row_ptr = A_row_ptr
         ctx.A_col_ind = A_col_ind
         ctx.sparse_structure = sparse_structure
```

Comment:
If I remember correctly, it is better to use `ctx.save_for_backward()`.

Reply:
This is very old code that Maurizio added. I'm going to refactor this in a future PR.